### PR TITLE
THREESCALE-10381 ProxyConfigPromote CR - improve config to stage promotion

### PR DESCRIPTION
**Pull request by ConorOM1**
# Issue Link

Jira: https://issues.redhat.com/browse/THREESCALE-10381

# What

Additional checks are required when bumping from Config environment to Staging environment via ProxyConfigPromote CR. These checks will prevent bumping if no Product changes are detected. The PCP CR will also reflect this action.

# Verification

- Provision a cluster
- Once ready, log in to your cluster via CLI.
- checkout this branch

- Install CRDs:
```
make install
```

- Create a new project:
```
oc new-project threescale
```

- Create a new secret:
```
oc apply -f - <<EOF   

kind: Secret
apiVersion: v1
metadata:
  name: s3-credentials
  namespace: threescale
data:
  AWS_ACCESS_KEY_ID: UkVQTEFDRV9NRQ==
  AWS_BUCKET: UkVQTEFDRV9NRQ==
  AWS_REGION: UkVQTEFDRV9NRQ==
  AWS_SECRET_ACCESS_KEY: UkVQTEFDRV9NRQ==
type: Opaque
EOF
```
- Create a new API Manager CR using your own Openshift console route as the wildcardDomain:
```
oc apply -f - <<EOF   

apiVersion: apps.3scale.net/v1alpha1
kind: APIManager
metadata:
  name: apimanager-sample
  namespace: threescale
spec:
  system:
    fileStorage:
      simpleStorageService:
        configurationSecretRef:
          name: s3-credentials
  wildcardDomain: apps.comalley.ip84.s1.devshift.org<CHANGE TO YOUR OWN OPENSHIFT CONSOLE ROUTE>
EOF
```
- Install dependencies:
```
make download
```

- Run the operator locally:
```
make run
```

- Wait until all resources are ready, you can check this with the following command:
```
oc get apimanager apimanager-sample -oyaml | yq '.status'
```

- Create a new backend CR:
```
oc apply -f - <<EOF   
apiVersion: capabilities.3scale.net/v1beta1
kind: Backend
metadata:
  name: backend1-cr
  namespace: threescale
spec:
  mappingRules: 
    - httpMethod: GET
      increment: 1
      last: true
      metricMethodRef: hits
      pattern: /
  name: backend1
  privateBaseURL: 'https://api.example.com/'
  systemName: backend1
EOF
```

- Create a new Product CR:
```
oc apply -f - <<EOF   
apiVersion: capabilities.3scale.net/v1beta1
kind: Product
metadata:
  name: product1-cr
  namespace: threescale
spec:
  name: product1
  backendUsages:
    backend1:
      path: /
EOF
```

- Log in to cluster UI and navigate to `Network` > then `Routes`
- Log in to 3scale portal via your `system-provider` link: 
    - user `admin` 
    - password: use `system-seed` password
- Once logged in, verify that `product1` is created
- Select `product1` and navigate to `Integration` > `Configuration`
- Verify that product is currently in Configuration environment and that Staging & Production are empty.

## Verification 1

- Create ProxyConfigPromote CR, ensure production is set to `false`:
```
oc apply -f - <<EOF   
apiVersion: capabilities.3scale.net/v1beta1
kind: ProxyConfigPromote
metadata:
  name: product1-v1-production
  namespace: threescale
spec:
  productCRName: product1-cr
  production: false
  deleteCR: true
EOF
```
- Expected outcome: Product will bump to the empty Staging environment. PCP CR will automatically delete.

## Verification 2

- Create the same PCP CR:
```
oc apply -f - <<EOF   
apiVersion: capabilities.3scale.net/v1beta1
kind: ProxyConfigPromote
metadata:
  name: product1-v1-production
  namespace: threescale
spec:
  productCRName: product1-cr
  production: false
  deleteCR: true
EOF
```
- Expected outcome: No changes detected so Product will not bump to Staging. PCP CR will remain with a notification.

## Verification 3

- Make a small change to the`Product CR` YAML. Save & Reload

- Monitor the same PCP CR:
```
oc apply -f - <<EOF   
apiVersion: capabilities.3scale.net/v1beta1
kind: ProxyConfigPromote
metadata:
  name: product1-v1-production
  namespace: threescale
spec:
  productCRName: product1-cr
  production: false
  deleteCR: true
EOF
```
- Expected outcome: Change is detected. This will trigger the new version bump to Staging. PCP CR will automatically delete.

## Verification 4

- Create another PCP CR. This time set production to `true`:
```
oc apply -f - <<EOF   
apiVersion: capabilities.3scale.net/v1beta1
kind: ProxyConfigPromote
metadata:
  name: product1-v1-production
  namespace: threescale
spec:
  productCRName: product1-cr
  production: true
  deleteCR: true
EOF
```
- Setting to `true` should trigger bump to Production environment. PCP CR will automatically delete.

### CODE REVIEW

File: controllers/capabilities/proxyconfigpromote_controller.go [hunk 1]
```diff
@@ -162,15 +181,13 @@ func (r *ProxyConfigPromoteReconciler) proxyConfigPromoteReconciler(proxyConfigP
 			}
 			latestStagingVersion = stageElement.ProxyConfig.Version
 
-			// Fetch production state, if product has not been promoted to production yet the state would be 0.
-			productionElement, err := threescaleAPIClient.GetLatestProxyConfig(productIDStr, "production")
```
Comment (MStokluska) [inline controllers/capabilities/proxyconfigpromote_controller.go:166]: I think this is still required.
Reason why is the following scenario:
- I have a configuration in stage (let's say version 4) and production (at version 3).
- I create a proxy config promote with stage bump to version 5.
- During reconciliation, if I pass the stage check and it turns out that both of my versions are different, the stage bump is successfull
- We are then moving to line 234 and sending in the latestStage and latestProduction version, in your approach, the production is always at 0 - ignored. 
- The reason why we sometimes might want to keep the Pcp CR is for tracking, and if so, a production version should also be set therefore we should not be sending nil value if it's not nil.

Comment (carlkyrillos) [reply controllers/capabilities/proxyconfigpromote_controller.go]: Done

**Comment by carlkyrillos** [comment]
@MStokluska This PR is ready to be re-reviewed if you have time

**Review by MStokluska — APPROVED** [review]

**Pull request merged**